app/Foliage/RemoteAsset.hs

```diff
@@ -50,11 +57,67 @@ addFetchRemoteAssetRule cacheDir = addBuiltinRule noLint noIdentity run
       let oldETag = fromMaybe BS.empty old
 
       newETag <-
-        withTempFile $ \fp -> traced "curl" $ do
-          BS.writeFile fp oldETag
-          createDirectoryIfMissing True (takeDirectory path)
-          cmd_ Shell ["curl", "--silent", "--location", "--etag-compare", fp, "--etag-save", fp, "--output", path, show uri]
-          BS.readFile fp
+        withTempFile $ \etagFile -> do
+          liftIO $ createDirectoryIfMissing True (takeDirectory path)
+          liftIO $ BS.writeFile etagFile oldETag
+          actionRetry 5 $ runCurl uri path etagFile
 
       let changed = if newETag == oldETag then ChangedRecomputeSame else ChangedRecomputeDiff
       return $ RunResult {runChanged = changed, runStore = newETag, runValue = path}
+
+runCurl :: URI -> String -> String -> Action ETag
+runCurl uri path etagFile = do
+  (Exit exitCode, Stdout out) <-
+    traced "curl" $
+      cmd
+        Shell
+        [ "curl",
+          -- Silent or quiet mode. Do not show progress meter or error messages. Makes Curl mute.
+          "--silent",
+          -- Fail fast with no output at all on server errors.
+          "--fail",
+          -- If the server reports that the requested page has moved to a different location this
+          -- option will make curl redo the request on the new place.
+          -- NOTE: This is needed because github always replies with a redirect
+          "--location",
+          -- This  option  makes  a conditional HTTP request for the specific ETag read from the
+          -- given file by sending a custom If-None-Match header using the stored ETag.
+          -- For correct results, make sure that the specified file contains only a single line
+          -- with the desired ETag. An empty file is parsed as an empty ETag.
+          "--etag-compare",
+          etagFile,
+          -- This option saves an HTTP ETag to the specified file. If no ETag is sent by the server,
+          -- an empty file is created.
+          "--etag-save",
+          etagFile,
+          -- Write output to <file> instead of stdout.
+          "--output",
+          path,
+          "--write-out",
+          "%{json}",
+          -- URL to fetch
+          show uri
+        ]
+  case exitCode of
+    ExitSuccess -> liftIO $ BS.readFile etagFile
+    ExitFailure c -> do
+      -- We show the curl exit code only if we cannot parse curl's write-out.
+      -- If we can parse it, we can craft a better error message.
+      case Aeson.eitherDecode out :: Either String CurlWriteOut of
+        Left err ->
+          error $
```

Comment:
`fail`? I guess clearly something is catching `ErrorCall`, but it seems nicer to use the failure that `Action` provides.

Reply:
(and there's another below)

Reply:
Totally agree with you on that.